quint/src/effects/inferrer.ts

```diff
@@ -63,73 +59,46 @@ export class EffectInferrer implements IRVisitor {
   private substitutions: Substitutions = []
 
   private builtinSignatures: Map<string, Signature> = getSignatures()
-  private lookupTable: LookupTableByModule
+  private lookupTable: LookupTable
   private freshVarGenerator: FreshVarGenerator
 
   // Track location descriptions for error tree traces
   private location: string = ''
 
-  private currentTable: LookupTable = newTable({})
-  private currentScopeTree: ScopeTree = { value: 0n, children: [] }
-
   enterExpr(e: QuintEx) {
     this.location = `Inferring effect for ${expressionToString(e)}`
   }
 
+  /* { kind: 'const', identifier: c } ∈ Γ
+   * ------------------------------------- (NAME-CONST)
+   *       Γ ⊢ c: Pure
+   */
+  exitConst(def: QuintConst) {
+    const effect: Effect = { kind: 'concrete', components: [] }
+    this.addToResults(def.id, right(toScheme(effect)))
+  }
+
+  /*  { kind: 'var', identifier: v } ∈ Γ
+   * ------------------------------------ (NAME-VAR)
+   *          Γ ⊢ v: Read[v]
+   */
+  exitVar(def: QuintVar) {
+    const effect: Effect = {
+      kind: 'concrete', components: [{ kind: 'read', entity: { kind: 'concrete', stateVariables: [{ name: def.name, reference: def.id }] } }],
+    }
+    this.addToResults(def.id, right(toScheme(effect)))
+  }
```

Comment:
I'm now adding the type/effects to the const/var definitions themselves and then name expressions only propagate these type/effects by using the lookup table. The implementation for `exitName` is significantly simplified for both type and effect inference.